content/en/blog/_posts/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm.md

```diff
@@ -0,0 +1,204 @@
+
+If you have no such opportunity to enter the chassis through iLO and collect a list of MAC addresses for your nodes, you can consider using [catchall cluster](https://asciinema.org/a/407287) pattern. Purely speaking, it is just a cluster with a dynamic DHCP pool. Thus, all nodes that are not described in the configuration to other clusters will automatically join to this cluster.
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot09.jpg)
+
+For example here we have a catchall cluster with some nodes. They are joined to the cluster with an auto-generated name based on their MAC address. We can connect to them and see what happens there. Here you can somehow prepare them, for example, set up the file system and then rejoin them to another cluster. 
+
+Now let's try connecting to the node terminal and see how it is booting. After the BIOS, the network card is configured, here it sends a request to the DHCP server from specific MAC address, which redirects it to a specific PXE server. Later the kernel and initrd image are downloaded from the server using the standard HTTP protocol:
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot10.jpg)
+
+After loading the kernel, the node downloads the rootfs image and transfers control to systemd. Then the booting proceeds as usual, and after that the node joins Kubernetes: 
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot11.jpg)
+
+If you take a look at **fstab**, you can see only two entries there: **/var/lib/docker** and **/var/lib/kubelet**, they are mounted as **tmpfs** (in fact, from RAM). At the same time, the root partition is mounted as **overlayfs**, so all changes that you make here on the system will be lost on the next reboot. 
+
+Looking into the block devices on the node, you can see some nvme disk, but it has not yet been mounted anywhere. There is also a loop device - this is the exactly rootfs image downloaded from the server. At the moment it is located in RAM, occupies 653 MB and mounted with the **loop** option.
+
+If we look in **/etc/ltsp**, we find our **network.sh** file that was executed at boot. From containers, we see running kube-proxy and pause container for it.
+
+![](/images/blog/2021-09-10-kubernetes-in-kubernetes-and-pxe-bootable-servers-farm/screenshot12.jpg)
+
+## Details
+
+### Network Boot Image
+
+But where does the main image come from? There is a little trick here. The image for the nodes is built through the [Dockerfile](https://github.com/kvaps/kubefarm/tree/master/build/ltsp) along with the server. The [Docker multi-stage build](https://docs.docker.com/develop/develop-images/multistage-build/) feature allows you to easily add any packages and kernel modules exactly at the stage of the image build. It looks like this:
```

Comment:
```suggestion
But where does the main image come from? there is a little trick here. The image for the nodes is built through the [Dockerfile](https://github.com/kvaps/kubefarm/tree/master/build/ltsp) along with the server. The [Docker multi-stage build](https://docs.docker.com/develop/develop-images/multistage-build/) feature allows you to easily add any packages and kernel modules exactly at the stage of the image build. It looks like this:
```

Reply:
@Mr-DG-Wick thank you for review!
Could you explain why is this correct, please?

Reply:
In general English grammar, Capitalization in the middle of a sentence is only permitted for the letter "i" that too only for certain words that form with the letter, like I, I'll, I'm etc. So I thought maybe you forgot to change that. 


Or maybe its some stylistic preference that I'm not aware of. So, you may or may not change that. Its **optional**

Reply:
Isn't it a new sentence?

Reply:
Hey. I just thinking is it common situation in English when question mark is not ends a sentence?

Reply:
see this @kvaps https://www.thepunctuationguide.com/terminal-points.html#:~:text=Rule%3A%20The%20question%20mark%20ends%20the%20sentence%3B%20no%20additional%20period.&text=Rule%3A%20The%20question%20mark%20(inside,sentence%3B%20no%20additional%20question%20mark.